samples/web/content/apprtc/gcm_register.py

```diff
@@ -0,0 +1,222 @@
+    msg = self.get_message_from_json()
+    if msg.has_key(PARAM_USER_ID_LIST) and msg[PARAM_USER_ID_LIST]:
+      result = []
+      for id in msg[PARAM_USER_ID_LIST]:
+        # TODO(jiayl): Only return the verified users when SMS verification is
+        # added.
+        if len(GCMRecord.get_by_user_id(id)) > 0:
+          result.append(id)
+      self.response.out.write(json.dumps(result))
+    elif msg.has_key(PARAM_GCM_ID) and msg[PARAM_GCM_ID]:
+      records = GCMRecord.get_by_gcm_id(msg[PARAM_GCM_ID])
+      assert len(records) < 2
+      if len(records) == 0:
+        self.response.out.write(constants.RESPONSE_NOT_FOUND)
+      elif records[0].verified:
+        self.response.out.write(constants.RESPONSE_VERIFIED)
+      else:
+        self.response.out.write(constants.RESPONSE_UNVERIFIED)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def post(self, cmd):
+    if cmd == 'new':
+      self.handle_new()
+    elif cmd == 'update':
+      self.handle_update()
+    elif cmd == 'verify':
+      self.handle_verify()
+    elif cmd == 'del':
+      self.handle_del()
+    elif cmd == 'query':
+      self.handle_query()
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+class GCMRecord(ndb.Model):
+  """Models a user ID to GCM ID binding record."""
+  user_id = ndb.StringProperty()
+  gcm_id = ndb.StringProperty()
+  code = ndb.StringProperty()
+  code_sent_time = ndb.DateTimeProperty()
+  verified = ndb.BooleanProperty()
+
+  @staticmethod
+  def get_by_user_id(user_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.user_id == user_id).fetch()
+
+  @staticmethod
+  def get_by_gcm_id(gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key())
+    return q.filter(GCMRecord.gcm_id == gcm_id).fetch()
+
+  @staticmethod
+  def has(user_id, gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    return len(q.fetch()) > 0
+
+  @staticmethod
+  @ndb.transactional(retries=100)
+  def add_or_update(user_id, gcm_id, code):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    records = q.fetch()
+    if len(records) == 0:
+      record = GCMRecord(parent = get_ancestor_key(),
+                         user_id = user_id,
+                         gcm_id = gcm_id,
+                         code = code,
+                         verified = False,
+                         code_sent_time = datetime.datetime.now())
+      record.put()
```

Comment:
Is the plan to have a cronjob that cleans old records out periodically? Or use TaskQueue with eta?